### ISSUE TITLE
Much better support needed for going between AMS and external spool

### ISSUE THREAD

**Issue by GuzziRaz** [bambulab/BambuStudio#3469]
**TL;DR**
A user with an AMS should not have four available types/colors of filament, but five (obviously with some manual interaction, but NOT requiring any G-code hacks). Ideally, *every aspect of the software, firmware and G-code should aim for that* (as far as reasonable). For example: We should be able to print a model that switches from PLA in the AMS to TPU 95A in the external spool, and back. No weird hacks, just do the intuitive things and print. When user interaction is need for the external spool, prompt for it and pause.

**Problems**
- Any user that doesn't have AMS, or that wants to print TPU 95A, will intuitively think that inserting a "Change Filament at layer" will do just that - but it's blatantly ignored without a warning, as discussed in #1958. BTW that issue title should be edited, this applies to the X1 series as well.
- When switching filament manually during a job, the brand, type and Flow Dynamics Calibration cannot be changed even for the external spool. It's greyed out. For certain use cases, that's a severe limitation.
- As soon as you switched to the external spool, there's no way back for the rest of the print job. This is a bad/sad limitation.

**Solutions**
- A "Change Filament" at layer can't ever be ignored without a as much as a pause/warning. The quite obvious way to handle it for the external spool is to automagically use `M600`. See additional context for a suggestion.
- During any pause, we should be able to switch to/from the external spool at will. If it was printing from the external spool, during the pause we should be able to manually unload that, select an AMS slot and hit Load, and then Resume.
- During a pause (in general, and for switching filament during print in particular) we should be able to reconfigure a spool (eg. changing brand/type and K-value profile).
- If possible, the AMS RFID scanning should work during a pause while no filament is loaded to the hotend.

Overall, there shouldn't be "AMS or not". There should be "AMS and external spool" at all times. The AMS mapping done when initiating a print should include the external spool as a choice (not by disabling AMS entirely, but by saying "Ext" for a particular filament)! Imagine a model that uses black, white and red. If I have black and white loaded in the AMS, and red loaded in the external spool, the mapping should default to just that. When printing: If the model starts with black and/or white, it uses the AMS. Once red is needed, a manual filament switch [to the external spool] is initiated, obviously with a manual pause for user intervention. If the reverse is needed, a manual pause is also done, for user to unload the external spool. Then they can press Resume and the printer will load next filament from the AMS.

**Alternatives/workarounds**
It appears inserting "Pause" as opposed to "Change Filament" for all filament switches, and using the external spool for the whole job, is what currently "works". There is no logic in that. Also, whatever Flow Dynamics Calibration is selected for the external spool at the start of job will presumably be used for the whole print - which can be totally wrong. Even during the pause, there is no way to change it.
As I wrote in https://github.com/bambulab/BambuStudio/issues/1958#issuecomment-1907895004, manually adding a pause in the top of Change Filament G-code can work around some problems (barely). But I suspect that too will print the PLA with the K-value from TPU, as it can't be changed during the filament switch...

**Additional context**
As a first step, there should be a boolean placeholder testable for "AMS enabled". It would be false if there is no AMS or if we disabled AMS when initiating the print. This way the Change Filament G-code could be written like this:
```gcode
{if ams_enabled}
    ; The current G-code for AMS goes here
    M620 S[next_extruder]A
    M204 S9000
      (... rest of current G-code...)
    M621 S[next_extruder]A
{else}
    ; Manual filament switch
    M600 [...]
{endif}
```

**Comment by andrewroth** [comment]
Almost a year to the day, I'm running into this kind of issue.

I prefer cura. So I right click the generated gcode, open with bambu studio, hit print. Disabled AMS.

M600 are ignored.
M400 U1 are also ignored, for some reason.

Pausing from the printer screen works. Why can't this be triggered from a gcode command?

Why do I feel I'm fighting against bambu's infrastructure instead of having it work for me?
